rust/connlib/tunnel/src/client.rs

```diff
@@ -223,9 +193,73 @@ pub struct ClientState {
     active_candidate_receivers: StreamMap<GatewayId, RTCIceCandidateInit>,
     /// We split the receivers of ICE candidates into two phases because we only want to start sending them once we've received an SDP from the gateway.
     waiting_for_sdp_from_gatway: HashMap<GatewayId, Receiver<RTCIceCandidateInit>>,
+
+    // TODO: Make private
+    pub awaiting_connection: HashMap<
+        ResourceId,
+        (
+            AwaitingConnectionDetails,
+            ResourceDescription,
+            Vec<GatewayId>,
+        ),
+    >,
+    pub gateway_awaiting_connection: HashMap<GatewayId, Vec<IpNetwork>>,
+
+    awaiting_connection_timers: StreamMap<ResourceId, Instant>,
+}
+
+#[derive(Debug, Clone, Copy, Default, PartialEq, Eq)]
+pub struct AwaitingConnectionDetails {
+    pub total_attemps: usize,
+    pub response_received: bool,
 }
 
 impl ClientState {
+    pub fn is_awaiting_connection_to(&self, resource: ResourceId) -> bool {
+        self.awaiting_connection.contains_key(&resource)
+    }
+
+    pub fn insert_new_awaiting_connection(
+        &mut self,
+        resource: ResourceDescription,
+        additional_gateways: impl Iterator<Item = GatewayId>,
+    ) {
+        const MAX_SIGNAL_CONNECTION_DELAY: Duration = Duration::from_secs(2);
+
+        let resource_id = resource.id();
+
+        let connected_gateway_ids = self
+            .gateway_awaiting_connection
+            .clone()
+            .into_keys()
+            .chain(additional_gateways)
+            .collect();
+
+        tracing::trace!(
+            gateways = ?connected_gateway_ids,
+            "connected_gateways"
+        );
+
+        self.awaiting_connection.insert(
+            resource_id,
+            (Default::default(), resource, connected_gateway_ids),
+        );
+
+        // TODO: Handle error
+        let _ = self.awaiting_connection_timers.try_push(
+            resource_id,
+            stream::poll_fn({
+                let mut interval = tokio::time::interval(MAX_SIGNAL_CONNECTION_DELAY);
+                move |cx| interval.poll_tick(cx).map(Some)
+            }),
+        );
+    }
+
+    pub fn remove_awaiting_connection(&mut self, resource: ResourceId) {
+        self.awaiting_connection.remove(&resource);
+        // self.awaiting_connection_timers.get_mut(&resource).cancel(); FIXME: Needs cancellation support.
```

Comment:
just as a note, this will only be used for error handling. If thinkgs are working normally, the timer will always be cleaned up before `awaiting_connection`, since that the latter shouldn't be cleaned up until there's an open data channel.